### PR TITLE
Implement ayon automator as the Build Control System

### PR DESCRIPTION
**Pull request by Lypsolon**
## Changelog Description

This PR adds the Ayon Automator and crates a few different Build Stages improving the CI and Developer Experience.

You can either set the variables as per the JSON Schema or use the recommended way with a build config YAML. 

`build_config.yaml`:
```yaml
# AyonUsd
AyonUsd_Linux: 
  COMPILEPLUGIN: "AyonUsdLinux/AyonUsd23_5_Py39_Linux"
  AyonUsdRoot: "$HOME/Documents/GitHub/Lyonh/ayon_usd_files/usd-24.03_linux_py39/24.03/"

AyonUsd_Windows: 
  COMPILEPLUGIN: "AyonUsdWin/AyonUsd23_5_Py39_Win"
  AyonUsdRoot: "C:\\Users\\ynput\\Documents\\GitRepos\\usd-24.03_win64_py39\\usd-24.03_win64_py39\\23.04"
```

The automator allows to define all build targets in the build config in the YAML format. 
Then you can use a Target option to specify what you want to build (Default is all).

_build conf Spec_
```
RUN_NAME: 
  COMPILEPLUGIN: "<Build Plugin Name"
  // List of Env Variables for the Build Plugin
  <Key>: "<Val>"
```

In the end all variables under the run name will be used as environment variables. 

## Additional info

The `AyonCppApi` uses the same system, but just not as complex. 

## Testing notes:

setup. 
when using the automator you need to run setup command before using the rest. 
1. python Project.py --setup

*dose the automator work*
1. python Project.py --help
the tool should in the terminal show the -- options. 
under runStageGRP and exec options there should be a list of things you can select from. 

*compile* 
1. python Project.py --runStageGRP "Generate Release" --setVars '{"BuildConf":"$HOME/Documents/build_conf.yaml"}'

the tool should compile all revolvers and then generate both a lakectl package called AyonUsdReslverBin inside of Revolvers and a zip with the same name. 

the build_conf.yaml can be named how ever you want but needs to be a yaml. 
this is my config. 

the software packages need to be installed and the path needs to be correct. 
if you can add or remove as you like. 

for ayon usd you can use the usd lib downloaded by the ayon usd addon. 
this is the simplest to test. 

```
# the name for the entry needs to have the target platform.system() Linux, Windows 

# AyonUsd
AyonUsd_Linux: 
  COMPILEPLUGIN: "AyonUsdLinux/AyonUsd23_5_Py39_Linux"
  AyonUsdRoot: "$HOME/Documents/GitHub/Lyonh/ayon_usd_files/usd-24.03_linux_py39/24.03/"

AyonUsd_Windows: 
  COMPILEPLUGIN: "AyonUsdWin/AyonUsd23_5_Py39_Win"
  AyonUsdRoot: "C:\\Users\\ynput\\Documents\\GitRepos\\usd-24.03_win64_py39\\usd-24.03_win64_py39\\23.04"


# Houdini
Houdini195_Py37_Windows:
  COMPILEPLUGIN: "HouWin/Houdini195_Py37_Win"
  HFS: "C:\\Program Files\\Side Effects Software\\Houdini 19.5.976"

Houdini195_Py37_Linux:
  COMPILEPLUGIN: "HouLinux/Houdini195_Py37_Linux"
  HFS: "$HOME/hou/Houdini19.5.773/"

Houdini195_Py39_Windows:
  COMPILEPLUGIN: "HouWin/Houdini195_Py39_Win"
  HFS: "C:\\Program Files\\Side Effects Software\\Houdini 19.5.805"

Houdini195_Py39_Linux:
  COMPILEPLUGIN: "HouLinux/Houdini195_Py39_Linux"
  HFS: "$HOME/hou/Houdini19.5.805/"

Houdini20_Py39_Windows:
  COMPILEPLUGIN: "HouWin/Houdini20_Py39_Win"
  HFS: "C:\\Program Files\\Side Effects Software\\Houdini 20.0.705"

Houdini20_Py39_Linux:
  COMPILEPLUGIN: "HouLinux/Houdini20_Py39_Linux"
  HFS: "$HOME/hou/Houdini20.0.724/"

Houdini20_Py310_Windows:
  COMPILEPLUGIN: "HouWin/Houdini20_Py310_Win"
  HFS: "C:\\Program Files\\Side Effects Software\\Houdini 20.0.590"

Houdini20_Py310_Linux:
  COMPILEPLUGIN: "HouLinux/Houdini20_Py310_Linux"
  HFS: "$HOME/hou/Houdini20.0.751/"


# Maya
Maya2025_Py311_Windows:
  COMPILEPLUGIN: "MayaWin/Maya2025_Py311_Win"
  MAYAPATH: "C:\\Program Files\\Autodesk\\Maya2025"
  MAYAUSDDEVKITPATH: r"C:\\Users\\ynput\\Documents\\Maya2025\\devkit"
  MAYAUSDPATH: "C:\\Program Files\\Autodesk\\MayaUSD\\Maya2025\\0.27.0"

Maya2025_Py311_Linux:
  COMPILEPLUGIN: "MayaLinux/Maya2025_Py311_Linux"
  MAYAPATH: "/usr/autodesk/maya2025"
  MAYAUSDDEVKITPATH: "$HOME/maya/devkits/2025_2/devkit/"
  MAYAUSDPATH: "/usr/autodesk/mayausd/maya2025/0.29.0_202406120659-f8eb6c1/mayausd"

Maya2024_2_Py310_Windows:
  COMPILEPLUGIN: "MayaWin/Maya2024_2_Py310_Win"
  MAYAPATH: "C:\\Program Files\\Autodesk\\Maya2024"
  MAYAUSDDEVKITPATH: r"C:\\Users\\ynput\\Documents\\devkit"
  MAYAUSDPATH: "C:\\Program Files\\Autodesk\\MayaUSD\\Maya2024\\0.25.0"

Maya2024_2_Py310_Linux:
  COMPILEPLUGIN: "MayaLinux/Maya2024_2_Py310_Linux"
  MAYAPATH: "/usr/autodesk/maya2024"
  MAYAUSDDEVKITPATH: "$HOME/maya/devkits/2024_2/devkit/"
  MAYAUSDPATH: "/usr/autodesk/mayausd/maya2024/0.25.0_202310160731-bbc8cc8/mayausd"

```

### CODE REVIEW

**Comment by BigRoy** [comment]
This is best reviewed by someone else - I know too little about this to say anything sensible. I did take a minute to just cleanup the PR description itself.

File: Project.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    default=20,
+    help="Set the minimum available memory required to initiate a task.",
+)
+
+cmd_args = AyonUsdResolverPRJ.setup_prj()
+import yaml
+import jsonschema
+from jsonschema import validate
+
+parrallel_limit = threading.Semaphore(cmd_args.Parallel)
```
Comment (antirotor) [inline Project.py:71]:
```suggestion
parallel_limit = threading.Semaphore(cmd_args.Parallel)
```

File: Project.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+            "patternProperties": {"^((?!COMPILEPLUGIN).)*$": {"type": "string"}},
+            "required": ["COMPILEPLUGIN"],
+            "additionalProperties": False,
+        }
+    },
+    "additionalProperties": False,
+}
+
+
+def remove_all_build_foulders():
```
Comment (antirotor) [inline Project.py:92]:
```suggestion
def remove_all_build_folders():
```

File: Project.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        os.getcwd(),
+        bin_package_path,
+        f"{str(compile_plugin)}_{platform.system()}_{platform.machine()}",
+    )
+
+    temp_dir_for_zip = os.path.abspath(
+        os.path.join(resolvers_dir, "temp", compile_plugin)
+    )
+    zip_inner_foulder_name = os.path.basename(zip_path)
+    zip_inner_foulder_path = os.path.join(temp_dir_for_zip, zip_inner_foulder_name)
```
Comment (antirotor) [inline Project.py:116]:
```suggestion
    zip_inner_folder_name = os.path.basename(zip_path)
    zip_inner_folder_path = os.path.join(temp_dir_for_zip, zip_inner_folder_name)
```

File: Project.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+    if os.path.exists(temp_dir_for_zip):
+        shutil.rmtree(temp_dir_for_zip)
+    os.makedirs(zip_inner_foulder_path)
```
Comment (antirotor) [inline Project.py:120]:
```suggestion
    os.makedirs(zip_inner_folder_path)
```

File: Project.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+    shutil.copytree(resolver_source_path, zip_inner_foulder_path, dirs_exist_ok=True)
```
Comment (antirotor) [inline Project.py:122]:
```suggestion
    shutil.copytree(resolver_source_path, zip_inner_folder_path, dirs_exist_ok=True)
```

File: Project.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        path = _get_os_path(val)
+        if os.path.exists(path):
+            val = path
+        test_env[key] = val
+    return test_env
+
+
+def cmake_compile(name, cmake_args, env_data, clean_build: bool, build_type):
+    try:
+        parrallel_limit.acquire()
```
Comment (antirotor) [inline Project.py:150]:
```suggestion
        parallel_limit.acquire()
```

File: Project.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+            f"-j{os.cpu_count()}",
+        )
+        Cmake.cmake_command(
+            AyonUsdResolverPRJ,
+            build_env,
+            "--install",
+            f"build/{name}",
+        )
+    finally:
+        parrallel_limit.release()
```
Comment (antirotor) [inline Project.py:182]:
```suggestion
        parallel_limit.release()
```

File: Project.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    if isinstance(build_targets, str):
+        conf_path = _get_os_path(build_targets)
+        if build_targets.endswith("yaml") and os.path.isfile(conf_path):
+            with open(conf_path, "r") as file:
+                build_config = yaml.safe_load(file)
+    try:
+
+        validate(instance=build_config, schema=build_config_schema)
+    except jsonschema.exceptions.ValidationError as e:
+        raise ValueError(f"Config not Valid {e}, data:{build_config}")
```
Comment (antirotor) [inline Project.py:197]:
```suggestion
        raise ValueError(f"Config not Valid {e}, data:{build_config}") from e
```

File: Project.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+
+def _cmake_multi_build(
+    build_targets: Union[str, dict],
+    cmake_conf: dict,
+    clean_build: bool = False,
+    clean_all: bool = False,
+    build_type=BuildTypes.Release,
+):
+    remove_all_build_foulders()
```
Comment (antirotor) [inline Project.py:209]:
```suggestion
    remove_all_build_folders()
```

File: Project.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    )
+    temp_dir = os.path.join(resolvers_dir, "temp")
+    if os.path.exists(temp_dir):
+        shutil.rmtree(temp_dir)
+    os.chdir(current_dir)
+
+
+zip_package_stage = Stage("Zip_Bin_Package")
+zip_package_stage.add_funcs(Func("zip", _zip_bin_package))
+zip_package_stage.addArtefactFoulder(
```
Comment (antirotor) [inline Project.py:306]:
```suggestion
zip_package_stage.addArtefactFolder(
```

File: Project.py
Comment (antirotor) [inline Project.py:1]: this could make use of some docstrings :)

**Review by antirotor — CHANGES_REQUESTED** [review]
some typos, not sure if I've got all in the suggestions

File: Project.py (hunk 10 above)
Comment (MustafaJafar) [reply Project.py]: this also requires to rename https://github.com/ynput/ayon-automator/blob/f323086ff197e89b794beb40732885869b84ffbd/AyonCiCd/Project.py#L508

File: Project.py
Comment (Lypsolon) [reply Project.py:1]: done.

**Comment by MustafaJafar** [comment]
This method has to be renamed to `addArtefactFolder` https://github.com/ynput/ayon-automator/blob/f323086ff197e89b794beb40732885869b84ffbd/AyonCiCd/Project.py#L508

--- 

when running the following command in cmd, I faced an error
```
python Project.py --runStageGRP "Build and Zip" --setVars {\"BuildConf\":\"build_conf.yaml\"}
```
it was solved by adding this argument to the following function `type=json.loads`, [reference](https://stackoverflow.com/questions/7625786/type-dict-in-argparse-add-argument)

https://github.com/ynput/ayon-automator/blob/a595cab5532b30df538cd328d2187a830a94e202/AyonCiCd/Project.py#L313-L318

and then change this line https://github.com/ynput/ayon-automator/blob/a595cab5532b30df538cd328d2187a830a94e202/AyonCiCd/Project.py#L23
to 
```
extra_env_var_dict = values
```

**Comment by MustafaJafar** [comment]
Also, here are my testing steps in cmd: 
```
git checkout 20-feature-implement-ayonautomater-for-build-and-testing
git submodule update --init --recursive
python Project.py --setup
# Create build_conf.yaml  manually, I don't know what to put inside it.
python Project.py --runStageGRP "Generate Release" --setVars {\"BuildConf\":\"build_conf.yaml\"}
```

**Comment by Lypsolon** [comment]
> traceback: <traceb

nice an windows error i love it. 
welp time to find out if there are windows and linux differences again. 
Manny thanks for testing.

**Comment by Lypsolon** [comment]
> Also, here are my testing steps in cmd: and, I 've `build_conf.yaml` file in the `ayon-usd-resolver` repo, I don't know what to put inside it.
> 
> ```
> git checkout 20-feature-implement-ayonautomater-for-build-and-testing
> git pull origin 20-feature-implement-ayonautomater-for-build-and-testing
> git submodule update --init --recursive
> python Project.py --setup
> python Project.py --runStageGRP "Generate Release" --setVars {\"BuildConf\":\"build_conf.yaml\"}
> ```

the build_conf.yaml needs to look like this.  

```
# the name for the entry needs to have the target platform.system() Linux, Windows 

# AyonUsd
AyonUsd_Linux: 
  COMPILEPLUGIN: "AyonUsdLinux/AyonUsd23_5_Py39_Linux"
  AyonUsdRoot: "$HOME/Documents/GitHub/Lyonh/ayon_usd_files/usd-24.03_linux_py39/24.03/"

AyonUsd_Windows: 
  COMPILEPLUGIN: "AyonUsdWin/AyonUsd23_5_Py39_Win"
  AyonUsdRoot: "C:\\Users\\ynput\\Documents\\GitRepos\\usd-24.03_win64_py39\\usd-24.03_win64_py39\\23.04"


..... more Target Configs 
```

you can just put ayon usd compile into it to limit the work that gets done (if you have multiple "targets" the system will compile them in parallel but with the same system so its not needed for testing )

for reference you yaml can look like this. and the AyonUsdRoot needs to point to a valid usd lib. an easy way to get one is to just run the ayon usd addon and point to the downloaded usd lib in there. (PS: the usd lib might be in a second sub foulder aka /ayon-usd/downloads/usdLibName/v24.08/ <- and here you find the actual lib. if there is no version number folder just ignore this text) 
```
AyonUsd_Windows: 
  COMPILEPLUGIN: "AyonUsdWin/AyonUsd23_5_Py39_Win"
  AyonUsdRoot: "C:\\Users\\ynput\\Documents\\GitRepos\\usd-24.03_win64_py39\\usd-24.03_win64_py39\\23.04"
```

File: .gitignore [hunk 11]
```diff
@@ -13,6 +13,7 @@ build/
 dist/
 Docs/docs/
 Resolvers/
+Ayon_Usd_Resolver_cicd/
```
Comment (MustafaJafar) [inline .gitignore:16]: What about gitignoring `build_conf.yaml`. I like to have it in my repo clone.
```suggestion
Ayon_Usd_Resolver_cicd/
build_conf.yaml
```

Comment (MustafaJafar) [reply .gitignore:16]: the conf file can have any name so it won't make much sense to have it like this.

**Comment by MustafaJafar** [comment]
I've pulled the latest commits and it started working. 
Here's the next error 😅 
```
----------------------------------- __del__() -----------------------------------
("Finished Execution with code:1, {'Cmake': ['Error at "
 'build/AyonUsd_Windows/src/AyonUsdResolver/cmake_install.cmake:55 (file):\\n  '
 'file INSTALL cannot find\\n  '
 '"E:/Ynput/ayon-usd-resolver/build/AyonUsd_Windows/src/AyonUsdResolver/Release/ayonUsdResolver.dll":\\n  '
 'File exists.\\nCall Stack (most recent call first):\\n  '
 "build/AyonUsd_Windows/cmake_install.cmake:47 (include)']}",)
--------------------------------------------------------------------------------

Exception ignored in: <function Project.__del__ at 0x000002154A9418B0>
Traceback (most recent call last):
  File "E:\Ynput\ayon-usd-resolver\automator\AyonCiCd\Project.py", line 193, in __del__
SystemExit: 1
```

**Comment by MustafaJafar** [comment]
Hey,
Here are step by step mini guide to test this PR. I think this guide can be useful for documenting how to build the resolver from [ayon-usd-resolver](https://github.com/ynput/ayon-usd-resolver) using the CI tools featured/enhanced in this PR.

## Intro
This PR adds/enhances the automation to build the available target within `Ayon_Usd_Resolver_cicd` folder. 
It'd be better to download the USD build from our [LakeFs Repo](https://lake.ayon.cloud). we provide two builds one for linux and one for windows. find them [here](https://lake.ayon.cloud/repositories/ayon-usd/objects?ref=main&path=AyonUsdBin%2F).

personally, I downloaded `usd-24.03_windows_py39.zip` as I'm on windows. 

## Testing Steps
1. Create your `build_conf.yaml`. you can name it what you want. personally, I added it in my `ayon-usd-resolver` clone
2. Specify desired targets to build in your `build_conf.yaml`. Personally, I'm going with `AyonUsd_Windows` and `Houdini195_Py39_Windows`
    ```yaml
    # AYON
    AyonUsd_Windows: 
    COMPILEPLUGIN: "AyonUsdWin/AyonUsd23_5_Py39_Win"
    AyonUsdRoot: "H:\\USDROOT\\usd-24.03_win64_py39\\23.04"  # path to the downloaded USD build.


    # Houdini
    Houdini195_Py39_Windows:
    COMPILEPLUGIN: "HouWin/Houdini195_Py39_Win"
    HFS: "C:\\Program Files\\Side Effects Software\\Houdini 19.5.805"  # path to my Houdini 19.5.805 install directory.
    ```
3. Checkout the correct branch, pull updates and update the submodules
    ```
    git checkout 20-feature-implement-ayonautomater-for-build-and-testing
    git pull origin 20-feature-implement-ayonautomater-for-build-and-testing
    git submodule update --init --recursive
    ```
4. Setup & Build
    ```
    python Project.py --setup
    python Project.py --runStageGRP "Generate Release" --setVars {\"BuildConf\":\"build_conf.yaml\"}
    ```

**Comment by Lypsolon** [comment]
> I've pulled the latest commits and it started working. Here's the next error 😅
> 
> ```
> ----------------------------------- __del__() -----------------------------------
> ("Finished Execution with code:1, {'Cmake': ['Error at "
>  'build/AyonUsd_Windows/src/AyonUsdResolver/cmake_install.cmake:55 (file):\\n  '
>  'file INSTALL cannot find\\n  '
>  '"E:/Ynput/ayon-usd-resolver/build/AyonUsd_Windows/src/AyonUsdResolver/Release/ayonUsdResolver.dll":\\n  '
>  'File exists.\\nCall Stack (most recent call first):\\n  '
>  "build/AyonUsd_Windows/cmake_install.cmake:47 (include)']}",)
> --------------------------------------------------------------------------------
> 
> Exception ignored in: <function Project.__del__ at 0x000002154A9418B0>
> Traceback (most recent call last):
>   File "E:\Ynput\ayon-usd-resolver\automator\AyonCiCd\Project.py", line 193, in __del__
> SystemExit: 1
> ```

yes many thanks. 
i got the error too and it was a change in the ayon cpp api i forgot an target include while changing to target cmake. 

i just updated it and it should work now i hope. 

if its not that can you upload the file Ayon_Usd_Resolver_cicd/artefacts/Build Release.txt ?

**Comment by MustafaJafar** [comment]
I've updated everything and I didn't hit any errors.
Here's my [Build Release.txt](https://github.com/user-attachments/files/17261023/Build.Release.txt)

Also, I can find the created zip package `ayon-usd-resolver\Ayon_Usd_Resolver_cicd\artefacts\Zip_Bin_Package_Artefacts\AyonUsdResolverBin.zip` which contains the two plugins specified in my `build_conf.yaml`.
image

**Review by antirotor — APPROVED** [review]

**Review by MustafaJafar — APPROVED** [review]
It works on windows.

**Comment by Lypsolon** [comment]
> It works on windows. For documentation purposes, here are the steps I followed to run it on windows [#36 (comment)](https://github.com/ynput/ayon-usd-resolver/pull/36#issuecomment-2393916204)

oh yeah i should fix the docs my bad. 
one moment

**Pull request merged**